Bitcoin Cash

DesWurstes opened the pull request:
As discussed

DesWurstes committed:
Tell mocha about tsconfig.json

DesWurstes committed:
Add rest api, networks, RPC fixes

DesWurstes committed:
fix signBatchP2SHTransaction parameters

DesWurstes commented on packages/bitcoin-cash-rest-api-provider/lib/BitcoinCashRestApiProvider.ts:
```diff
@@ -0,0 +1,197 @@
+      version: rawTx['version'],
+      locktime: rawTx['locktime'],
+      vin: rawTx['vin'],
+      fee: 0, // TODO
```
This is not a "will do" but rather a "don't care" that I justified in the last section of [this](https://github.com/liquality/chainabstractionlayer/pull/405#issue-630515718). Simply, costs outweigh no use for this number.

monokh commented on packages/bitcoin-cash-rest-api-provider/lib/BitcoinCashRestApiProvider.ts:
```diff
@@ -0,0 +1,197 @@
+import { ChainProvider, Address, bitcoin, BigNumber } from '@liquality/types'
+import * as explorer from './types'
+import { BitcoinCashNetwork } from '@liquality/bitcoin-networks'
+import BCHJS from '@psf/bch-js'
```
If I understand correctly, this package simply makes rest calls. 
Throughout the library, we have avoided using rest clients to avoid introducing an extra dependency (security).
Any chance these calls can be made with axios?

monokh commented on packages/bitcoin-cash-rest-api-provider/lib/BitcoinCashRestApiProvider.ts:
```diff
@@ -0,0 +1,197 @@
+  }
+
+  async getFeePerByte(numberOfBlocks = this._numberOfBlockConfirmation) {
+    numberOfBlocks as any
```
What does this line do?

DesWurstes replied:
Used to silence a warning

monokh commented on packages/bitcoin-cash-rest-api-provider/lib/BitcoinCashRestApiProvider.ts:
```diff
@@ -0,0 +1,197 @@
+
+  async getUnspentTransactions(_addresses: (Address | string)[]): Promise<bitcoin.UTXO[]> {
+    const addresses = _addresses.map(addressToString)
+    const utxoSets = await Promise.all(addresses.map((addr) => this._getUnspentTransactions(addr)))
```
This fullstack api [seems to support querying multiple addresses](https://api.fullstack.cash/docs/#api-ElectrumX_/_Fulcrum-UTXOs_for_an_array_of_addresses) in one query. Could we use that for the efficiency gain? We do similarly on the bitcoin side

monokh commented on packages/bitcoin-cash-rest-api-provider/lib/BitcoinCashRestApiProvider.ts:
```diff
@@ -0,0 +1,197 @@
+    return transactions['transactions'].length
+  }
+
+  async getAddressTransactionCounts(_addresses: (Address | string)[]) {
```
Any batch api available here similar to the utxo one?

monokh commented on packages/bitcoin-cash-rest-api-provider/lib/types.ts:
```diff
@@ -0,0 +1,78 @@
+export type FeeEstimates = { [index: string]: number }
```
This seems identical to the `BitcoinEsploraApiProvider` types
In that case it should be moved to the types package, in the bitcoin folder i guess

monokh commented on packages/bitcoin-js-wallet-provider/lib/BitcoinJsWalletProvider.ts:
```diff
@@ -26,7 +34,11 @@ export default class BitcoinJsWalletProvider extends BitcoinWalletProvider(
+    super({
+      network,
+      baseDerivationPath,
+      addressType: network.segwitCapable ? addressType : bitcoin.AddressType.LEGACY
```
This is making it a bit unclear IMO, if the parameter passed is incorrect then it shouldn't be silently changed

An alternative can be to throw an error if `segwitCapable` is false and the wrong address type is passed

Users of the library should pass legacy when using BCH

monokh commented on packages/bitcoin-js-wallet-provider/lib/BitcoinJsWalletProvider.ts:
```diff
@@ -156,22 +208,59 @@ export default class BitcoinJsWalletProvider extends BitcoinWalletProvider(
+        const sigHash = tx.hashForWitnessV0(index, inputs[i].outputScript, inputs[i].vout.vSat, 0x01 | 0x40)
+        const signed = keyPairs[i].sign(sigHash)
+
+        // BitcoinJS does not allow SIGHASH_FORKID for encoding
```
Could you potentially extract this function and add some comments about what it does, perhaps a source for the code?

monokh commented on packages/bitcoin-js-wallet-provider/lib/BitcoinJsWalletProvider.ts:
```diff
@@ -156,22 +208,59 @@ export default class BitcoinJsWalletProvider extends BitcoinWalletProvider(
     for (let i = 0; i < inputs.length; i++) {
       const index = inputs[i].txInputIndex ? inputs[i].txInputIndex : inputs[i].index
+
+      if (this._network.protocolType == ProtocolType.BitcoinCash) {
```
Is it specific to bitcoin cash or is there more generality to it? If other forks have the same process, should call the parameter `ProtocolType.BitcoinCash` something more generic

monokh replied:
not sure if this sounds right, but maybe there can be a `forkId` parameter in the network ?

monokh commented on packages/bitcoin-js-wallet-provider/lib/BitcoinJsWalletProvider.ts:
```diff
@@ -64,6 +76,41 @@ export default class BitcoinJsWalletProvider extends BitcoinWalletProvider(
+      })
+      inputAddressInfo.push(inputs[i].address)
+    }
+    const sigs = await this.signBatchP2SHTransaction(inputInfo as any, inputAddressInfo as any, tx, 0)
```
Why the `as any`? Parameters seem compatible

monokh commented on packages/bitcoin-networks/lib/index.ts:
```diff
@@ -30,4 +70,158 @@ const BitcoinNetworks = {
+
+// Bitcoin Cash
+
+function cashaddrToLegacy(address: string) {
```
We like to keep the BitcoinNetwork class serializable as it gets passed between contexts sometimes

Is this format only for bitcoin cash? In that case it would be great if it could be moved to `bitcoin-utils` instead. 

something like 

```
addToBitcoinJs (address: string, network: BitcoinNetwork) {
 if(!bitcoincash) return address

  ....
 // handle the cash address format
}
```

monokh commented on packages/bitcoin-swap-provider/lib/BitcoinSwapProvider.ts:
```diff
@@ -37,7 +37,7 @@ export default class BitcoinSwapProvider extends Provider implements Partial<Swa
     }
     this._network = network
-    this._mode = mode
+    this._mode = network.segwitCapable ? mode : bitcoin.SwapMode.P2SH
```
Similar to the other point about throwing an error if the wrong type was supplied

DesWurstes commented:
> This seems identical to the BitcoinEsploraApiProvider types
In that case it should be moved to the types package, in the bitcoin folder i guess

Fixed, moved it to types. Please make sure to update calling CAL from other libs because this is a breaking change: esplora-api-provider no longer exports esplora, and it's..... **Never mind I moved it now in a way that the previous API still exists in a working condition**

> Is it specific to bitcoin cash or is there more generality to it? If other forks have the same process, should call the parameter ProtocolType.BitcoinCash something more generic
@monokh monokh 20 hours ago Member

> not sure if this sounds right, but maybe there can be a forkId parameter in the network ?

This is specific to BCH, BCHABC, BCHSV and BTG. BTG does not even sign the same way BCH does, while BCHABC and BCHSV both use the same forkId as BCH, so I don't see it as worth implementing. BCHABC, BCHSV fit perfectly in the current BitcoinCashNetwork struct, but if BTG were to be added, then it would be closer to BitcoinNetwork which would need forkId. Simply, I don't think we should add that.

I have nothing to add, but further reviews and ideas still welcome :)

monokh commented on packages/bitcoin-esplora-swap-find-provider/lib/BitcoinEsploraSwapFindProvider.ts:
```diff
@@ -10,18 +10,28 @@ type PaymentVariants = {
 
 export default class BitcoinEsploraSwapFindProvider extends NodeProvider {
-  constructor(url: string) {
+  _useRest: boolean
```
it looks like you are trying to support 2 apis with the same code, technically all the requests are rest calls
would it be better to create another provider that extends this and implements the `findAddressTransaction` differently?

DesWurstes commented:
This batch provider turned out to need to use the Ninsight API. The BCHJS package that used to be there was actually hiding that this was using a separate API. Done.

DesWurstes committed:
BCH Swap find provider

This reverts commit 92c1bd4d360abf2c2144d4d837a8797816075106.

DesWurstes committed:
BCH tests remove duplicate code

DesWurstes committed:
Diff that needs manual rebasing
